tycho-gpg-plugin/src/main/java/org/eclipse/tycho/gpg/SignRepositoryArtifactsMojo.java

```diff
@@ -30,213 +26,269 @@
+    enum PGPKeyBehavior {
+        skip, replace, merge
+    }
 
     /**
      * The maven project.
      */
     @Parameter(defaultValue = "${project}", readonly = true, required = true)
     protected MavenProject project;
 
+    /**
+     * The repository location.
+     */
     @Parameter(defaultValue = "${project.build.directory}/repository")
     private File repository;
 
     /**
-     * Configures to <code>true</code> to generate PGP signature only for artifacts that do
-     * <strong>not</strong> already contain signatures files from jarsigner.
+     * Configure to {@code true} to generate PGP signature only for artifacts that are
+     * <strong>not</strong> signed by a jarsigner.
      */
     @Parameter(defaultValue = "true")
     private boolean skipIfJarsigned;
 
     /**
-     * If set to {@code true} adds the public key belonging to the signature to the repository
+     * Configure to {@code true} to generate a PGP signature only for artifacts that do
+     * <strong>not</strong> contain a signature that's anchored in Java's trust store, i.e.,
+     * anchored in the JDK's cacerts. A JCA certificate, for example, is never anchored.
+     */
+    @Parameter(defaultValue = "true")
+    private boolean skipIfJarsignedAndAnchored;
+
+    /**
+     * Configure to {@code true} to generate a PGP signature for binary artifacts.
+     */
+    @Parameter(defaultValue = "true")
+    boolean skipBinaries;
+
+    /**
+     * Configure to {@code true} to add the public key of each signature to the repository's
      * metadata.
      */
     @Parameter(defaultValue = "true")
     private boolean addPublicKeyToRepo;
 
     /**
-     * If set to {@code true} adds the public key belonging to the signature to each signed
-     * artifact's metadata.
+     * Configure to {@code true} to add the public key of the signature to each signed artifact's
+     * metadata.
      */
     @Parameter(defaultValue = "true", alias = "addPublicKeysToArtifacts")
     private boolean addPublicKeyToArtifacts;
 
     /**
-     * Bundles that should be signed independently of other settings, e.g. {@link #skipIfJarsigned}.
+     * Configures how to generate PGP signatures for artifacts that already have one or more PGP
+     * signatures, {@code skip} to generate no new PGP signature, {@code replace} to replace the
+     * existing signature(s) with a new signature, and {@code merge} to add a new signature to any
+     * existing signature(s).
+     */
+    @Parameter(defaultValue = "skip")
+    private PGPKeyBehavior pgpKeyBehavior;
+
+    /**
+     * Configure the signer used for PGP signing. Currently supported are {@code gpg} for launching
+     * the native {@code gpg} executable, and {@code bc} for using Bouncy Castle libraries. The
+     * latter is much faster and it can sign in parallel, so is very much faster.
+     */
+    @Parameter(defaultValue = "pgp")
+    private String signer;
```

Comment:
Are you sure.  I think that's risky because a great many options are ignored in this case and also note that even the bouncy castle thing only works because we can get the gpg keys, so without gpg one just cannot get secret keys from anywhere.

Reply:
The property thing is a good idea though.  What name do you suggest for it?

Reply:
I changed it to "gpg" until you are really sure.  I added the property pgp.signer.

Reply:
We usually use something prefixed with "tycho" so probably `tycho.pgp.signer`.

As long as users can "switch back" with a configuration option it is fine, and i think BC is the future and should be improved/promoted as much as possible, so what do you think is missing? Without checking it in details I would expect not much more than specify a secret key (maybe a passphrase?) on the commandline or the maven config and then everything is signed with that key. There seem [some examples on how to accomplish it with BC](https://github.com/rodbate/bouncycastle-examples/blob/master/src/main/java/bcfipsin100/pgp/Sign.java), then it could be enhanced to probably read some of the "magic locations" where usually stuff is stored to automate this more, and we probabbly want to reuse the maven credentials encryption later on, but thats of course all optional for a first step.

So if you could use that in your test-setup I think it should be usable in "real code" as well?

Reply:
I will change it to tycho.pgp.signer for the next force push.

I will further investigate if the one use of gpg can be eliminated by reading the underlying secrete keys and the public keystore directly; I already have code for the latter.  This would have the advantage that one doesn't have to configure or start the gpg agent making it significantly easier to set up a build.  But let's do that as a separate step and if that's successful then we can change the default to bc.   I.e., I'd very much like to get past this PR because I expect I'm going to need it (the ability to sign features) to deal with the SimRel 2022-12 signing problems.  

Okay?

Reply:
Agree, lets go for it!